presto-main/src/main/java/com/facebook/presto/operator/scalar/RowToRowCast.java

```diff
@@ -0,0 +1,163 @@
+import java.util.List;
+
+import static com.facebook.presto.bytecode.Access.FINAL;
+import static com.facebook.presto.bytecode.Access.PUBLIC;
+import static com.facebook.presto.bytecode.Access.STATIC;
+import static com.facebook.presto.bytecode.Access.a;
+import static com.facebook.presto.bytecode.CompilerUtils.defineClass;
+import static com.facebook.presto.bytecode.Parameter.arg;
+import static com.facebook.presto.bytecode.ParameterizedType.type;
+import static com.facebook.presto.bytecode.expression.BytecodeExpressions.constantBoolean;
+import static com.facebook.presto.bytecode.expression.BytecodeExpressions.constantInt;
+import static com.facebook.presto.bytecode.expression.BytecodeExpressions.newInstance;
+import static com.facebook.presto.metadata.OperatorType.CAST;
+import static com.facebook.presto.metadata.Signature.internalOperator;
+import static com.facebook.presto.metadata.Signature.withVariadicBound;
+import static com.facebook.presto.sql.gen.InvokeFunctionBytecodeExpression.invokeFunction;
+import static com.facebook.presto.sql.gen.SqlTypeBytecodeExpression.constantType;
+import static com.facebook.presto.util.Reflection.methodHandle;
+import static com.google.common.base.Preconditions.checkArgument;
+
+public class RowToRowCast
+        extends SqlOperator
+{
+    public static final RowToRowCast ROW_TO_ROW_CAST = new RowToRowCast();
+
+    private RowToRowCast()
+    {
+        super(CAST, ImmutableList.of(withVariadicBound("F", "row"), withVariadicBound("T", "row")), ImmutableList.of(), "T", ImmutableList.of("F"));
+    }
+
+    @Override
+    public ScalarFunctionImplementation specialize(BoundVariables boundVariables, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        checkArgument(arity == 1, "Expected arity to be 1");
+        Type fromType = boundVariables.getTypeVariable("F");
+        Type toType = boundVariables.getTypeVariable("T");
+        if (fromType.getTypeParameters().size() != toType.getTypeParameters().size()) {
+            throw new PrestoException(StandardErrorCode.INVALID_FUNCTION_ARGUMENT, "the size of fromType and toType must match");
+        }
+        Class<?> castOperatorClass = generateRowCast(fromType, toType, functionRegistry);
+        MethodHandle methodHandle = methodHandle(castOperatorClass, "castRow", ConnectorSession.class, Block.class);
+        return new ScalarFunctionImplementation(false, ImmutableList.of(false), methodHandle, isDeterministic());
+    }
+
+    private static Class<?> generateRowCast(Type fromType, Type toType, FunctionRegistry functionRegistry)
+    {
+        List<Type> toTypes = toType.getTypeParameters();
+        List<Type> fromTypes = fromType.getTypeParameters();
+
+        CallSiteBinder binder = new CallSiteBinder();
+
+        ClassDefinition definition = new ClassDefinition(
+                a(PUBLIC, FINAL),
+                CompilerUtils.makeClassName(Joiner.on("$").join("RowCast", fromType, toType)),
```

Comment:
Also, this may fail if `fromType.toString()` or `toType.toString()` contain characters that are not allowed in class names.